src/main.ts

```diff
@@ -7,479 +7,1026 @@ import {
-      const hotkey = this.plugin.settings.hotkeys[i];
-      if (
-        hotkey.key !== existingHotkey.key ||
-        hotkey.meta !== existingHotkey.meta ||
-        hotkey.shift !== existingHotkey.shift
-      ) {
-        continue;
-      }
+  private readonly loadSavedSettings = async (): Promise<void> => {
+    writeConsole('Loading previously saved settings.');
 
-      console.debug(
-        `Updating leader-hotkey command ${hotkeyToName(
-          existingHotkey,
-        )} at index ${i} to ${newCommand}`,
+    const savedSettings = (await this.loadData()) || {};
+    try {
+      writeConsole('Loaded previous settings.');
+      savedSettings.hotkeys = (savedSettings.hotkeys || []).map(KeyMap.of);
+      this.settings = savedSettings;
+    } catch (err) {
+      writeConsole('A failure occured while parsing the saved settings.');
+      createNotice(
+        'A failure occured while loading the saved settings. Fallbacking to defaults.',
       );
-      hotkey.commandID = newCommand;
-      break;
-    }
-    this.plugin.saveData(this.plugin.settings);
-  };
+      // todo : Retrocompatibility?
+      //  Harder than i thought since LeaderKey isn't saved here.
+      //  Would need to keep the old command ,
+      //  lookup the binding and convert it to the new one.
 
-  private readonly storeNewHotkeyInSettings = (): void => {
-    console.debug(
-      `Adding leader-hotkey command ${this.tempNewHotkey} to ${this.tempNewHotkey.commandID}`,
-    );
-    this.plugin.settings.hotkeys.push(this.tempNewHotkey);
-    this.plugin.saveData(this.plugin.settings);
-    this.tempNewHotkey = newEmptyHotkey();
+      this.settings = defaultSettings;
+    }
+    this.matchHandler = new MatchHandler(this);
   };
 }
 
-const newEmptyHotkey = (): Hotkey => ({
-  key: '',
-  shift: false,
-  meta: false,
-  commandID: '',
-});
-
-const hotkeyToName = (hotkey: Hotkey): string => {
-  if (hotkey === undefined || hotkey.key === '') {
-    return '?';
-  }
-  const keyToUse = (() => {
-    switch (hotkey.key) {
-      case 'ArrowRight':
-        return '→';
-      case 'ArrowLeft':
-        return '←';
-      case 'ArrowDown':
-        return '↓';
-      case 'ArrowUp':
-        return '↑';
-      default:
-        return hotkey.key;
-    }
-  })();
-  return (
-    (hotkey.meta ? 'meta+' : '') + (hotkey.shift ? 'shift+' : '') + keyToUse
-  );
+// region consts and utils
+const listCommands = (app: App): ObsidianCommand[] => {
+  // todo remove any type
+  const anyApp = app as any;
+  const commands = anyApp.commands.commands as CommandMap;
+  return Object.values(commands);
+};
+const classifyMatch = (bestMatch: Optional<TrieNode<KeyMap>>): MatchKind => {
+  if (!bestMatch) {
+    return MatchKind.NoMatch;
+  }
+  if (bestMatch.isLeaf()) {
+    return MatchKind.FullMatch;
+  }
+  return MatchKind.PartialMatch;
+};
+const defaultHotkeys: KeyMap[] = [
+  new KeyMap('editor:focus-left', [KeyPress.ctrl('b'), KeyPress.just('h')]),
```

Comment:
Do these defaults work? Or does the live preview steal them? If they don't we should pick a new default leader.

Reply:
They do work, assuming no conflicting keybinds. The 'proper' ones, mapped via the native plugin seem to take precedence. In these cases, i don't even receive the keyboard event.